### PR TITLE
Select polling based watcher for docker log file watcher on Windows

### PR DESCRIPTION
**Pull request by mnltejaswini**
Signed-off-by: Tejaswini Duggaraju <naduggar@microsoft.com> <br/> <br/>

- **Fixes** #30046 <br/><br/>

- **What I did**  <br/> <br/>

I updated logger to use polling based watcher for Windows .  This is because the file watcher based on syscall notifications doesn't work because of file caching.  <br/><br/>

- **How I did it**  <br/><br/>

Based on runtime OS, I selected poller based watcher for Windows. <br/><br/>

- **How to verify it** <br/><br/>

`docker run -it --name <containername>  <container image> ` <br/>
`docker logs -f <containername>`  <br/><br/>

This should stream the logs of the container continuously. <br/><br/>

- **Description for the changelog** <br/><br/>
Check for the runtime OS and use poller based watcher to work around the file caching issue in Windows <br/>

### CODE REVIEW

File: daemon/logger/loggerutils/logfile.go [hunk 1]
```diff
@@ -641,6 +642,15 @@ func followLogs(f *os.File, logWatcher *logger.LogWatcher, notifyRotate chan int
 }
 
 func watchFile(name string) (filenotify.FileWatcher, error) {
+
+	// FileWatcher on Windows files is based on the syscall notifications which has an issue becuase of file caching.
+	// It is based on ReadDirectoryChangesW() which doesn't detect writes to the cache. It detects writes to disk only.
+	// Becuase of the OS lazy writing, we don't get notifications for file writes and thereby the watcher
+	// doesn't work. Hence for Windows we will use poll based notifier.
+	if runtime.GOOS == "windows" {
+		return watchFileForWindows(name)
```
Comment (cpuguy83) [inline daemon/logger/loggerutils/logfile.go]: This ends up duplicating a lot.
Can we set `fileWatcher` to a polling watcher for Windows and keep using the rest of this function?

**Review by cpuguy83 — CHANGES_REQUESTED** [review]
Thanks!

File: daemon/logger/loggerutils/logfile.go [hunk 2]
```diff
@@ -641,26 +642,42 @@ func followLogs(f *os.File, logWatcher *logger.LogWatcher, notifyRotate chan int
...
+		// doesn't work. Hence for Windows we will use poll based notifier.
+		fileWatcher = filenotify.NewPollingWatcher()
+	} else {
+		var err error
+		fileWatcher, err = filenotify.New()
+		if err != nil {
+			return nil, err
+		}
+
+		if err := fileWatcher.Add(name); err == nil {
```
Comment (cpuguy83) [inline daemon/logger/loggerutils/logfile.go]: Can probably keep this check outside the if-block. Not particularly harmful to error out twice in the case of Windows.

Comment (mnltejaswini) [reply daemon/logger/loggerutils/logfile.go]: Not sure I followed this comment. For non Windows we retry with polling right?

Comment (mnltejaswini) [reply daemon/logger/loggerutils/logfile.go]: Wouldn't the fallback logic be duplicate for Windows?

Comment (cpuguy83) [reply daemon/logger/loggerutils/logfile.go]: Yes, but doesn't really hurt anything.

Comment (mnltejaswini) [reply daemon/logger/loggerutils/logfile.go]: Ok. I updated the code as per your comment.

**Review by cpuguy83 — APPROVED** [review]
LGTM

**Comment by mnltejaswini** [comment]
The test case that failed is 
FAIL: check_test.go:347: DockerSwarmSuite.TearDownTest
18:54:13 
18:54:13 unmount of /tmp/docker-execroot/d5970d64d43e6/netns failed: invalid argument
18:54:13 unmount of /tmp/docker-execroot/d3cd72178e45a/netns failed: no such file or directory
18:54:13 unmount of /tmp/docker-execroot/dd97a72566d92/netns failed: invalid argument
18:54:13 unmount of /tmp/docker-execroot/d333a093c501c/netns failed: no such file or directory
18:54:13 check_test.go:352:
18:54:13     d.Stop(c)
18:54:13 /go/src/github.com/docker/docker/internal/test/daemon/daemon.go:401:
18:54:13     t.Fatalf("Error while stopping the daemon %s : %v", d.id, err)
18:54:13 ... Error: Error while stopping the daemon db5c2a7fd6e84 : exit status 130
18:54:13 
18:54:13 
18:54:13 ----------------------------------------------------------------------
18:54:13 PANIC: docker_api_swarm_test.go:780: DockerSwarmSuite.TestAPISwarmRestartCluster

My change is not directly related to it. Is it expected or do I need to investigate?

**Comment by cpuguy83** [comment]
Unrelated testing bug.

**Review by boaz0 — APPROVED** [review]
LGTM :+1:

**Review by thaJeztah — APPROVED** [review]
LGTM

ping @johnstep @jhowardmsft

**Comment by jhowardmsft** [comment]
LGTM

**Pull request merged**